### PR TITLE
Fix/2094/all files are rejected on import

### PR DESCRIPTION
**Pull request by IhsenBouallegue**
# Improve file validation handling

Close: #2094 

## Description

  - All files used to be rejected on import if one is erroneous.
  - This PR enables all correct files (also ones containing just warnings) to be loaded, regardless of the erroneous files.
  - This PR implements better and more descriptive error messages on file import

## Screenshots or gifs

grafik

### CODE REVIEW

File: visualization/app/codeCharta/util/fileValidator.ts [hunk 1]
```diff
@@ -35,41 +35,57 @@ export const ERROR_MESSAGES = {
 export function validate(nameDataPair: NameDataPair) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/fileValidator.ts:35]: I do not think `validate()` needs to be exported.

File: visualization/app/codeCharta/util/fileValidator.ts [hunk 2]
```diff
@@ -35,41 +35,57 @@ export const ERROR_MESSAGES = {
...
 	const file = nameDataPair?.content
 	const result: CCValidationResult = { error: [], warning: [] }
+
+	result.error = checkErrors(file)
+	result.warning = checkWarnings(file)
+
+	if (result.error.length > 0) {
+		throw result.error
+	}
+}
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/fileValidator.ts:45]: I propably would call `checkErrors` and `checkWarnings` in loadFiles directly. The `validate` function is not needed anymore. And instead of throwing errors and warnings from one function to another, I would just process the return values of the two new check functions.

File: visualization/app/codeCharta/util/fileValidator.ts [hunk 3]
```diff
@@ -35,41 +35,57 @@ export const ERROR_MESSAGES = {
...
+
+function checkWarnings(file: ExportCCFile) {
+	const warnings = []
+	switch (true) {
+		case fileHasHigherMinorVersion(file):
+			warnings.push(`${ERROR_MESSAGES.minorApiVersionOutdated} Found: ${file.apiVersion}`)
+			break
+	}
+	return warnings
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/fileValidator.ts:54]:
```suggestion
	const warnings = []
	if (fileHasHigherMinorVersion(file)) {
			warnings.push(`${ERROR_MESSAGES.minorApiVersionOutdated} Found: ${file.apiVersion}`)
	}
	return warnings
```

File: visualization/app/codeCharta/util/fileValidator.ts [hunk 4]
```diff
@@ -35,41 +35,57 @@ export const ERROR_MESSAGES = {
...
-			result.error.push(ERROR_MESSAGES.majorApiVersionIsOutdated)
-			break
-		case fileHasHigherMinorVersion(file):
-			result.warning.push(`${ERROR_MESSAGES.minorApiVersionOutdated} Found: ${file.apiVersion}`)
+			errors.push(ERROR_MESSAGES.majorApiVersionIsOutdated)
 			break
 	}
 
-	if (result.error.length === 0) {
+	if (errors.length === 0) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/fileValidator.ts:71]: The JSON Schema validations is a separate concern and should therefore be extracted into a separate function.

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 2 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts:45]: Awesome that would simplify the code even further!

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 3 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts:54]: Will implement in the next commit.

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 1 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts]: Implemented in the next commit.

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 4 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts:65]: Implemented in the next commit.

File: visualization/app/codeCharta/util/fileValidator.ts [hunk 5]
```diff
@@ -32,42 +32,50 @@ export const ERROR_MESSAGES = {
...
 			break
 	}
+	if (errors.length === 0) errors.push(...checkJsonSchema(file))
+	return errors
+}
 
-	if (result.error.length === 0) {
+function checkJsonSchema(file: ExportCCFile) {
+	const errors: string[] = []
+	if (errors.length === 0) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/util/fileValidator.ts:65]: This is always true, isn't it?

File: visualization/app/codeCharta/codeCharta.service.ts [hunk 6]
```diff
@@ -25,28 +25,26 @@ export class CodeChartaService {
...
+
 		for (const nameDataPair of nameDataPairs) {
-			try {
-				validate(nameDataPair)
+			errors.push(...checkErrors(nameDataPair.content))
+			if (errors.length > 0) {
+				this.fileStates.filter(file => getCCFile(nameDataPair) !== file.file)
+				this.recentFiles.filter(fileName => fileName !== nameDataPair.fileName)
+			}
+			if (errors.length === 0) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/codeCharta.service.ts:37]: The behavior before was different. Errors and warnings were shown at the same time. Now, Warnings are shown only if no errors are present.

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 5 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts:65]: Missed that! I will clean it up, thanks 😃

File: visualization/app/codeCharta/codeCharta.service.ts (hunk 6 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/codeCharta.service.ts:37]: That is true. But it didn't make sense to me. Let's say a file has an error, which means something is seriously wrong with it. Why should we bother checking for warnings at this point? Plus, in the future if we decide to add warnings that assume that the file structure is correct, the old implementation is not flexible enough.
When we show the user the validation results, I think if a file has an error, the user won't care about its warnings. But once the error is fixed, the warnings will be checked for and shown if found.

Basically, the warnings of a certain file will be shown only when it has no errors, meaning the structure is correct. 
What do you think?

File: visualization/app/codeCharta/util/fileValidator.ts (hunk 5 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/util/fileValidator.ts:66]: Fixed in next commit

File: visualization/app/codeCharta/codeCharta.service.ts [hunk 7]
```diff
@@ -25,29 +25,19 @@ export class CodeChartaService {
...
+
+		this.getValidationResults(nameDataPairs, fileValidationResults)
 
-		if (this.fileStates.length > 0) {
+		this.storeService.dispatch(setIsLoadingFile(false))
+
+		if (
+			fileValidationResults.flatMap(validation => validation.errors).length > 0 ||
+			fileValidationResults.flatMap(validation => validation.warnings).length > 0
+		)
```
Comment (ce-bo) [inline visualization/app/codeCharta/codeCharta.service.ts:37]: It would move line 30 into getValidationResults(). Thus, you could just return an empty array (on no errors) or a filled one with errors or warnings. Line 34-37 would then just be a check on `fileValidationResults.length > 0` without the two flatMaps.

File: visualization/app/codeCharta/codeCharta.service.ts [hunk 8]
```diff
@@ -25,29 +25,19 @@ export class CodeChartaService {
...
+			await this.dialogService.showValidationDialog(fileValidationResults)
+
+		if (this.recentFiles.length > 0) {
```
Comment (ce-bo) [inline visualization/app/codeCharta/codeCharta.service.ts:40]: It looks like that recentFiles is not set anymore,

File: visualization/app/codeCharta/ui/dialog/dialog.service.ts [hunk 9]
```diff
@@ -38,26 +38,45 @@ export class DialogService {
...
+
+		if (fileValidationResults.flatMap(validation => validation.errors).length > 0) {
+			htmlMessages.push("<h2>Errors</h2>")
+			for (const fileValidationResult of fileValidationResults) {
+				if (fileValidationResult.errors.length > 0) {
+					const fileErrorMessage = this.buildFileErrorMessage(fileValidationResult)
+					htmlMessages.push(fileErrorMessage)
+				}
+			}
+		}
```
Comment (ce-bo) [inline visualization/app/codeCharta/ui/dialog/dialog.service.ts:52]: I would not iterate twice over the validation results. Here a suggestion to save one iteration:
```suggestion
		const errors = fileValidationResults.filter((validationResult) => { return validationResult.errors.length > 0 })
		if (errors .length > 0) {
			htmlMessages.push("<h2>Errors</h2>")
			for (const error of errors ) {
					const fileErrorMessage = this.buildFileErrorMessage(error )
					htmlMessages.push(fileErrorMessage)
			}
		}
```

**Review by ce-bo — COMMENTED** [review]
Would you please fix the two failing quality gates and check my last comments? :-)

**Comment by IhsenBouallegue** [comment]
> Would you please fix the two failing quality gates and check my last comments? :-)

I am working right now on the tests. Not my strong suit. But as soon as I get everything green, I will look into the comments! :D

File: visualization/app/codeCharta/codeCharta.service.ts (hunk 8 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/codeCharta.service.ts:40]: it is being set in method addFile

File: visualization/app/codeCharta/ui/dialog/dialog.service.ts (hunk 9 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/ui/dialog/dialog.service.ts:52]: Doesn't filter iterate over the array too? Or is it less expansive than flatmap?

Comment (ce-bo) [reply visualization/app/codeCharta/ui/dialog/dialog.service.ts:52]: Yes, it does. But your second loop in line 46 will iterate all validation results again (instead of just the errors). Furthermore, it must be checked if there are any errors present a second time  .

File: visualization/app/codeCharta/codeCharta.service.ts (hunk 7 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/codeCharta.service.ts:37]: Resolved in the next commit

File: visualization/app/codeCharta/ui/dialog/dialog.service.ts (hunk 9 above)
Comment (IhsenBouallegue) [reply visualization/app/codeCharta/ui/dialog/dialog.service.ts:52]: Resolved in the next commit

**Review by ce-bo — APPROVED** [review]
Code LGTM! :-)
